### PR TITLE
Refactor SQL/BigQuery/Qubole/Druid Check operators

### LINKED ISSUE #10271: BigQueryValueCheckOperator derived from deprecated ValueCheckOperator
`airflow.contrib.operators.bigquery_check_operator.BigQueryValueCheckOperator` [is derived from](https://github.com/apache/airflow/blob/69cc8370e486e00dd526e96c2b0aacd348ee255e/airflow/contrib/operators/bigquery_check_operator.py#L82) `airflow.operators.check_operator.ValueCheckOperator` but needs to be derived from `airflow.operators.sql.SQLValueCheckOperator` instead.

Until this is done, airflow generates a lot of warnings from [this](https://github.com/apache/airflow/blob/69cc8370e486e00dd526e96c2b0aacd348ee255e/airflow/operators/check_operator.py#L81) line of code.

### PR DESCRIPTION
**Pull request by xinbinhuang**
closes: #10271
related: #9844 #14184

This PR refactor SQL/BigQuery Check operators to reduce duplicated code:
- create `BaseSQLOperator`: it standardizes how some of the generic SQL operators retrieve DB hook with the `.get_db_hook()` method
  - Add a `database` kwarg `*CheckOperators` for a consistent interface 
- create `_BigQueryDbHookMixin` to standardize the `.get_db_hook()` method for BigQuery
- create `_QuboleCheckOperatorMixin` to remove duplicate code
- replace `<class-name>.template_fields` with `_get_template_fields` in `__getattribute__` to avoid hard coding class name, and reduce duplicate code
- remove and deprecate `DruidCheckOperator` the same functionality can be achieved by `SQLCheckOperator` - the deprecation method is the same for `PrestoCheckOperator`

Misc:
- Fix docstrings
- Update deprecated Operator name and import path
- Remove unnecessary `if` statements check parameters in `SQLBranchOperator`

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/providers/google/cloud/operators/bigquery.py [hunk 1]
```diff
@@ -87,7 +87,17 @@ def get_link(self, operator: BaseOperator, dttm: datetime):
...
+    def get_db_hook(self) -> BigQueryHook:
+        return BigQueryHook(
+            gcp_conn_id=self.gcp_conn_id,
+            use_legacy_sql=self.use_legacy_sql,
+            location=self.location,
+            impersonation_chain=self.impersonation_chain,
+        )
+
+
+class BigQueryCheckOperator(CheckOperator, BigQueryDbHookMixin):
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/bigquery.py:100]: This is not a pattern recommended by Google:
https://s.apache.org/google-operators

It is rather preferred to instantiated the hook in execute method in operator. In this way refactoring and future changes are simpler and we reduce cross dependencies between classes.

File: tests/operators/test_sql.py [hunk 2]
```diff
@@ -322,7 +359,7 @@ def test_pass_min_value_max_sql(self, mock_get_db_hook):
 
         operator.execute()
 
-    @mock.patch.object(ThresholdCheckOperator, "get_db_hook")
+    @mock.patch.object(SQLThresholdCheckOperator, "get_db_hook")
```
Comment (turbaszek) [inline tests/operators/test_sql.py:354]: Was this name changed?

**Comment by xinbinhuang** [comment]
@turbaszek thanks for taking a look! I submitted the PR pretty late last night so I didn't  have time to add description to the PR. I will update the description later today and double check the CI failure.

File: tests/operators/test_sql.py (hunk 2 above)
Comment (xinbinhuang) [reply tests/operators/test_sql.py:354]: Yeah. The original operators were moved to another module and renamed but the tests were forgotten to refactor I believe.

Comment (xinbinhuang) [reply tests/operators/test_sql.py:354]: You can see the original imports and the refactored imports at the top of the file - `test_sql.py`

Comment (turbaszek) [reply tests/operators/test_sql.py:354]: > tests were forgotten to refactor I believe.

That was what I thought, wanted to check. Thanks!

**Comment by turbaszek** [comment]
Hey @xinbinhuang there seems to be some pylint issues. Can you take a look?

**Comment by xinbinhuang** [comment]
@turbaszek  there is a pylint error that is not relevant to this PR. Do you want me to take a look at that too?

**Comment by turbaszek** [comment]
> @turbaszek there is a pylint error that is not relevant to this PR. Do you want me to take a look at that too?

Please take a look at the issue as I think it's related to your change - you introduced a new level of inheritance.

**Comment by xinbinhuang** [comment]
> > @turbaszek there is a pylint error that is not relevant to this PR. Do you want me to take a look at that too?
> 
> Please take a look at the issue as I think it's related to your change - you introduced a new level of inheritance.

Ahhh, you are right. Thanks for the tips. Let me fix it in a bit

File: airflow/providers/qubole/operators/qubole_check.py [hunk 3]
```diff
@@ -16,17 +16,40 @@
...
+
+    # this overwrite the original QuboleOperator.get_hook() which returns a QuboleHook.
+    def get_hook(self) -> QuboleCheckHook:
+        """Reinitialising the hook, as some template fields might have changed"""
+        return QuboleCheckHook(
+            context=self._hook_context, results_parser_callable=self.results_parser_callable, **self.kwargs
+        )
+
+
+# pylint: disable=too-many-ancestors
```
Comment (xinbinhuang) [inline airflow/providers/qubole/operators/qubole_check.py:51]: Temporarily disable pylint for these 2 operators. Will submit a second PR in the future to simplify the inheritance.

File: airflow/operators/sql.py [hunk 4]
```diff
@@ -38,7 +41,53 @@
...
+        self.conn_id = conn_id
+        self.database = database
+
+    @cached_property
+    def _hook(self):
+        """Get DB Hook based on connection type"""
+        self.log.debug("Get connection for %s", self.conn_id)
+        conn = BaseHook.get_connection(self.conn_id)
+
+        if conn.conn_type not in ALLOWED_CONN_TYPE:
```
Comment (potiuk) [inline airflow/operators/sql.py]: Hmm. I think we should rather check the povider's manager hook property for that. We've added automated discovery of providers so that they can register their own connection -> Hook mapping so maybe we should utilise that? https://github.com/apache/airflow/blob/6f246b0d54ccaf733b7c5951a8955adda6719acb/airflow/providers_manager.py#L387

File: airflow/providers/qubole/operators/qubole_check.py (hunk 3 above)
Comment (potiuk) [reply airflow/providers/qubole/operators/qubole_check.py:51]: Yeah. for now it is good, but maybe leave a TODO(xinbinhuang)  note as comment?

**Review by potiuk — COMMENTED** [review]
:heart:  it

Just a few small comment :)

**Comment by RosterIn** [comment]
@xinbinhuang shouldn't [DruidCheckOperator](https://github.com/apache/airflow/blob/master/airflow/providers/apache/druid/operators/druid_check.py#L26) be also covered here?

File: airflow/operators/sql.py (hunk 4 above)
Comment (xinbinhuang) [reply airflow/operators/sql.py]: I think the reason for `ALLOWED_CONN_TYPE` is to only whitelist a subset of all connection types that are `DbApiHook` compatible. Maybe we can do something like this instead?


```python
hook = conn.get_hook()
if not isinstance(hook, DbApiHook):
    raise AirflowException('The connection type associated hook is not a subclass of `DbApiHook`') 
```

However, there are edge cases that need to be resolved in future PRs to make this work.
- Currently, bigQuery seems to be not discoverable because it overlaps with the GoogleBaseHook: https://github.com/apache/airflow/blob/5f81fc73c8ea3fc1c3b08080f439fa123926f250/airflow/providers/google/cloud/hooks/bigquery.py#L72
- CloudSQLHook/CloudSQLDatabaseHook is not a subclass of `DbApiHook`
https://github.com/apache/airflow/blob/5f81fc73c8ea3fc1c3b08080f439fa123926f250/airflow/providers/google/cloud/hooks/cloud_sql.py#L666
- ... not sure if there are other edge cases

File: airflow/providers/apache/druid/operators/druid_check.py [hunk 5]
```diff
@@ -15,74 +15,23 @@
...
-    stop the critical path, preventing from
-    publishing dubious data, or on the side and receive email alerts
-    without stopping the progress of the DAG.
-
-    :param sql: the sql to be executed
-    :type sql: str
-    :param druid_broker_conn_id: reference to the druid broker
-    :type druid_broker_conn_id: str
+    This class is deprecated.
+    Please use `airflow.operators.sql.SQLCheckOperator`.
```
Comment (xinbinhuang) [inline airflow/providers/apache/druid/operators/druid_check.py:26]: Is there a deprecation process for Provider packages? i.e. any changes to `provider.yaml` and the README file?

@potiuk

Comment (potiuk) [reply airflow/providers/apache/druid/operators/druid_check.py:26]: There will be very shortly - I am working on adding SemVER versioning changes and procedure for providers.

File: airflow/providers/qubole/operators/qubole.py [hunk 6]
```diff
@@ -267,7 +267,7 @@ def get_hook(self) -> QuboleHook:
         return QuboleHook(**self.kwargs)
 
     def __getattribute__(self, name: str) -> str:
-        if name in QuboleOperator.template_fields:
+        if name in _get_template_fields(self):
```
Comment (msumit) [inline airflow/providers/qubole/operators/qubole.py:270]: what is the rationale behind this change?

Comment (xinbinhuang) [reply airflow/providers/qubole/operators/qubole.py:270]: It avoids hardcoding the class name `QuboleOperator` here, so you can inherit it safely. - so I can remove the `<class-name>.template_fields` in `qubole_check.py`

File: airflow/providers/apache/druid/operators/druid_check.py (hunk 5 above)
Comment (xinbinhuang) [reply airflow/providers/apache/druid/operators/druid_check.py:26]: So I will just keep the deprecation inside the `class`  for now. Do you think that is ok?

**Comment by xinbinhuang** [comment]
@turbaszek @potiuk Hi do you have any further feedback or comments that would want me to address?

File: airflow/operators/sql.py [hunk 7]
```diff
@@ -18,27 +18,59 @@
...
+
+        hook = conn.get_hook()
+        if not isinstance(hook, DbApiHook):
+            raise AirflowException(
+                f'The connection type is not supported by {self.__class__.__name__}. '
+                f'The associated hook should be a subclass of `DbApiHook`. Got {hook.__class__.__name__}'
+            )
+
+        if self.database:
+            hook.schema = self.database
```
Comment (turbaszek) [inline airflow/operators/sql.py:59]: This is slightly confusing, database is not the same thing as schema

Comment (xinbinhuang) [reply airflow/operators/sql.py:59]: I totally agree!  But this is how the current mapping of Airflow works. Currently, the `Connection` object uses the field `schema` to specify the database (not schema) for database type connections, i.e. Postgres, MySQL, etc. And as a result, all `DbApiHook`s use the field `schema` to specify the database to connect to using the underlying DB API (i.e. `psycopg2`). `extra` is used to specify the schema if applicable. So the relationship looks like this:

```
Connection(schema) -> DbApiHook (schema) -> DB API (dbname/database) 
```

About the field `database`, it was already used for the `SQLBranchOperator` and also other SQL like operators. For this PR, I simply refactor it out to reduce duplicated code and add them to the `check*` operators for consistency.

Though I think it will be nice to consolidate between `schema` and `database`, I think they should be addressed in a separate PR or worth a discussion in the `dev` list. I can send an email in the dev list if you don't object.

**Review by turbaszek — APPROVED** [review]

**Comment by turbaszek** [comment]
@xinbinhuang nice work! Just to be sure - there is no breaking change, right? @potiuk would you mind also taking a look?

**Comment by xinbinhuang** [comment]
> @xinbinhuang nice work! Just to be sure - there is no breaking change, right? @potiuk would you mind also taking a look?

Yes, there is one place that may be considered fixing rather than breaking. For `airflow/providers/google/cloud/hooks/bigquery.py`, I changed the hook's `conn_type` and `hook_name`  from `google cloud platform`/`Google Cloud` to `bigquery`/`BigQuery`. This fixes the issue where `conn.get_hook` will never return the bigquery hook because `BaseGcpHook` "hides" the `BigQueryHook` as they have the same `conn_type`.

**Comment by xinbinhuang** [comment]
Hi @potiuk, just a gentle ping.

Do you think the PR is safe to merge? Or are there other feedback or tests that you would want me to take a look?

**Comment by potiuk** [comment]
Hey @xinbinhuang . I am working on relasing providers and the "maintenance" process of them  - the PR and discussion is in #13767 I hope to get it merged tomorrow/Monday and relase a new wave of providers early next week. I think I'd prefer to add this one afterwards - I would love to take a close look at the provider's backwards - compatibility then and we could work together to test the new provider mechanism (and updates to semver versions of the providers that it brings) with you. 

Woudl that be OK ?

**Comment by xinbinhuang** [comment]
> Hey @xinbinhuang . I am working on releasing providers and the "maintenance" process of them - the PR and discussion is in #13767 I hope to get it merged tomorrow/Monday and release a new wave of providers early next week. I think I'd prefer to add this one afterwards - I would love to take a close look at the provider's backwards - compatibility then and we could work together to test the new provider mechanism (and updates to SEMVER versions of the providers that it brings) with you.
> 
> Would that be OK ?

Sounds great to me! I will try to chime in a bit in that PR when I have time.

**Comment by xinbinhuang** [comment]
Hi @potiuk , I noticed that #13767 has been merged. I wonder would be your suggestion on upgrading this PR to follow the versioning policy? Can you give me some pointers? Thanks!

**Comment by xinbinhuang** [comment]
@potiuk I increased the `apache.druid` provider version to `1.0.2` because of deprecating a backward-compatible operator. It seems that I can also generate the documentation and releases with `./breeze prepare-provider-packages` and `./breeze prepare-provider-documentation` but I am not sure if I should do these. Can you help take a look and advise on anything else that you would want me to add?

**Comment by xinbinhuang** [comment]
@kaxil @turbaszek @potiuk  Can we merge this as it's already approved? It's pretty hard to get CI happy because it often fails for reasons that not related to this PR.

**Comment by kaxil** [comment]
It is failing unit tests:

https://github.com/apache/airflow/pull/12677/checks?check_run_id=1932643565#step:6:942

```
  _ TestMovingCoreToContrib.test_is_class_deprecated_097_airflow_operators_sql_SQLCheckOperator _

  >           raise AssertionError(_error_message()) from cause
  E           AssertionError: expected call not found.
  E           Expected: __init__()
  E           Actual: __init__(conn_id='druid_broker_default')
  
  /usr/local/lib/python3.8/unittest/mock.py:913: AssertionError
  ________ TestMovingCoreToContrib.test_no_redirect_to_deprecated_classes ________


....


              if new in all_classes_by_old:
  >               raise AssertionError(
                      f'Deprecation "{old}" to "{new}" is incorrect. '
                      f'Please use \"{all_classes_by_old[new]}\" instead of "{old}".'
                  )
  E               AssertionError: Deprecation "airflow.operators.druid_check_operator.DruidCheckOperator" to "airflow.providers.apache.druid.operators.druid_check.DruidCheckOperator" is incorrect. Please use "airflow.operators.sql.SQLCheckOperator" instead of "airflow.operators.druid_check_operator.DruidCheckOperator".
  
  tests/core/test_core_to_contrib.py:116: AssertionError
  _ TestMovingCoreToContrib.test_warning_on_import_243_airflow_operators_sql_SQLCheckOperator _
```

File: tests/deprecated_classes.py [hunk 8]
```diff
@@ -1008,7 +1008,7 @@
         'airflow.contrib.operators.sqoop_operator.SqoopOperator',
     ),
     (
-        'airflow.providers.apache.druid.operators.druid_check.DruidCheckOperator',
+        'airflow.operators.sql.SQLCheckOperator',
```
Comment (xinbinhuang) [inline tests/deprecated_classes.py:1011]: Deprecate `DruidCheckOperator` in favor of `SQLCheckOperator`

File: tests/core/test_core_to_contrib.py [hunk 9]
```diff
@@ -84,7 +84,7 @@ def test_is_class_deprecated(self, new_module, old_module):
                     # is used (=3 or =4 if @apply_defaults)
                     assert len(warnings) >= 1
                     assert any(warning.filename == __file__ for warning in warnings)
-                init_mock.assert_called_once_with()
+                init_mock.assert_called_once()
```
Comment (xinbinhuang) [inline tests/core/test_core_to_contrib.py:87]: Not sure if there is a better way to do this while maintaining backward compatibility.  `DruidCheckOperator` has a default `druid_broker_conn_id` that needs to pass down to `SQLCheckOperator.__init__(conn_id=druid_broker_conn_id, **kwargs)`.

So if we do `assert_called_once_with()` the test will fail because `DruidCheckOperator.__ini__()` == `SQLCheckOperator.__init__(conn_id=druid_broker_conn_id)`

File: airflow/providers/apache/druid/operators/druid_check.py [hunk 10]
```diff
@@ -15,74 +15,22 @@
...
-            raise AirflowException("The query returned None")
-        self.log.info("Success.")
+    def __init__(self, druid_broker_conn_id: str = 'druid_broker_default', **kwargs):
+        warnings.warn(
+            """This class is deprecated.
+            Please use `airflow.operators.sql.SQLCheckOperator`.""",
+            DeprecationWarning,
+            stacklevel=3,
+        )
+        super().__init__(conn_id=druid_broker_conn_id, **kwargs)
```
Comment (xinbinhuang) [inline airflow/providers/apache/druid/operators/druid_check.py:36]: Passdown `druid_broker_conn_id` to `SQLCheckOperator.__init__`

**Review by xinbinhuang — COMMENTED** [review]
@kaxil  Thanks for taking a look! I made a small update to the `DruidCheckOperator` in `deprecated_classes.py` after the original comment. I didn't know that it will break the tests, but now it's fixed.

File: tests/core/test_core_to_contrib.py (hunk 9 above)
Comment (turbaszek) [reply tests/core/test_core_to_contrib.py:87]: @xinbinhuang should this help?
```
assert_called_once_with(conn_id=mock.ANY)
```

**Comment by turbaszek** [comment]
@ashb @mik-laj @potiuk would you mind taking a look?

**Review by kaxil — APPROVED** [review]

File: airflow/operators/druid_check_operator.py [hunk 11]
```diff
@@ -23,7 +23,7 @@
 from airflow.providers.apache.druid.operators.druid_check import DruidCheckOperator  # noqa
 
 warnings.warn(
-    "This module is deprecated. Please use `airflow.providers.apache.druid.operators.druid_check`.",
+    "This module is deprecated. Please use `airflow.operators.sql.SQLCheckOperator`.",
```
Comment (kaxil) [inline airflow/operators/druid_check_operator.py:26]: Oh why is this?

Comment (kaxil) [reply airflow/operators/druid_check_operator.py:26]: nvm saw the change later in the PR

File: airflow/providers/google/cloud/hooks/bigquery.py [hunk 12]
```diff
@@ -69,8 +69,8 @@ class BigQueryHook(GoogleBaseHook, DbApiHook):
 
     conn_name_attr = 'gcp_conn_id'
     default_conn_name = 'google_cloud_default'
-    conn_type = 'google_cloud_platform'
-    hook_name = 'Google Cloud'
+    conn_type = 'bigquery'
+    hook_name = 'BigQuery'
```
Comment (kaxil) [inline airflow/providers/google/cloud/hooks/bigquery.py]: @turbaszek @potiuk @mik-laj is this change ok?

Comment (kaxil) [reply airflow/providers/google/cloud/hooks/bigquery.py]: @xinbinhuang Do we need this change?

Comment (xinbinhuang) [reply airflow/providers/google/cloud/hooks/bigquery.py]: I left a comment about this in the thread previously, but I think it got buried over time. 

This fixes the issue that `conn.get_hook` will never return the bigquery hook after we start to use the `ProvidersManager` because GoogleBaseHook "hides" the BigQueryHook as they have the same conn_type.

Here I copy my original comment:

> Yes, there is one place that may be considered fixing rather than breaking. For airflow/providers/google/cloud/hooks/bigquery.py, I changed the hook's conn_type and hook_name from google cloud platform/Google Cloud to bigquery/BigQuery. This fixes the issue where conn.get_hook will never return the bigquery hook because GoogleBaseHook "hides" the BigQueryHook as they have the same conn_type.

Comment (kaxil) [reply airflow/providers/google/cloud/hooks/bigquery.py]: Does it needs any action from a user side is why I am thinking about it ? -- if yes, then we might need a note somewhere in Up:

https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/ADDITIONAL_INFO.md

Comment (xinbinhuang) [reply airflow/providers/google/cloud/hooks/bigquery.py]: As similar to other Google cloud connection type (before this change, you can't find bigquery in the list)

image
image

Maybe I should rename it to `Google Bigquery` instead?

Comment (xinbinhuang) [reply airflow/providers/google/cloud/hooks/bigquery.py]: Renamed, this is the new conn_type list for google:
image

Comment (xinbinhuang) [reply airflow/providers/google/cloud/hooks/bigquery.py]: > Does it needs any action from a user side is why I am thinking about it ? -- if yes, then we might need a note somewhere in Up: https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/ADDITIONAL_INFO.md

Hmm, I think we need a note for it. From sole user perspective, the conn_type for BigQuery changes from `Google Cloud` -> `Google BigQuery`. However, if we don't change it the current master branch will break or make it impossible on things that rely on `Connection.get_hook` to return a `BigQueryHook`.

Should I update the providers' version by 0.0.1 and add a note to the `ADDITIONAL_INFO.md` or other documentation?

Comment (kaxil) [reply airflow/providers/google/cloud/hooks/bigquery.py]: Yes probably add an Updating.md style note in `ADDITIONAL_INFO.md` with version as header -- same as https://github.com/apache/airflow/blob/master/UPDATING.md

Comment (kaxil) [reply airflow/providers/google/cloud/hooks/bigquery.py]: or in https://github.com/apache/airflow/blob/master/airflow/providers/google/CHANGELOG.rst actually

Comment (potiuk) [reply airflow/providers/google/cloud/hooks/bigquery.py]: Just a comment on this one. Yeah. I wanted to avoid this change when we switched to 2.0 (not to add backwards incompatibilities), but this change indeed breaks the compatibility of Google Provider, so we should mention that in the CHANGELOG @kaxil mentioned. I think however, whether this is a good time to "merge" this change now - we are planning a bigger release of Google Provider (3.0.0) around March/April when we will drop "backports" altogether. We need to update dependencies and upgrade a few remaining APIs of python libraries of Google - and such a backwards-incompatible change could be added then. 

I think if we could make this change without this breaking compatibility now, that would be ideal.

Comment (xinbinhuang) [reply airflow/providers/google/cloud/hooks/bigquery.py]: @potiuk I want to clarify a bit about the change here. The current Goold Providers in the master branch is broken because it doesn't return  `BigQueryHook` when we do `Connection.get_hook` (this is the original behavior). 

Before these 2 PRs #12466 #12558, `Connection.get_hook` was using a static `conn_type` -> `HookClass` mapping to get the underlying hook. However, after #12558, `airflow.providers.google.common.hooks.base_google.GoogleBaseHook` was added after `BigQueryHook` so `BigQueryHook` will be overridden during providers discovery. https://github.com/PolideaInternal/airflow/blob/802159767baf1768d92c6047c2fdb2094ee7a2a8/airflow/providers/google/provider.yaml#L714-L720

To solve this, I think we can either:
- Fix the bug by putting  `BigQueryHook`  after `GoogleBaseHook`. This will give back us the old behavior.
- Rename the `conn_type` as in this PR, which will involve some versioning consideration.

I think we can fix the bug in this PR, and create a new issue/PR with the new `conn_type`. I am not familiar with the overall releasing schedule and planning, so please let me know which one do you prefer? @potiuk @kaxil

Comment (xinbinhuang) [reply airflow/providers/google/cloud/hooks/bigquery.py]: I have reverted the `BigQueryHook` conn_type change and put `BigQueryHook` after `GoogleBaseHook` for now.

File: airflow/operators/sql.py [hunk 13]
```diff
@@ -120,16 +148,16 @@ def _convert_to_float_if_possible(s):
     return ret
 
 
-class SQLValueCheckOperator(BaseOperator):
+class SQLValueCheckOperator(BaseSQLOperator):
```
Comment (kaxil) [inline airflow/operators/sql.py:151]: Since this operator is in Airflow Core, can we release providers with the changes in this PR without corresponding core changes

**Comment by xinbinhuang** [comment]
@potiuk @kaxil  CI greenish except the MySQL transient failure...

File: airflow/operators/sql.py (hunk 13 above)
Comment (kaxil) [reply airflow/operators/sql.py:151]: cc @potiuk

Comment (xinbinhuang) [reply airflow/operators/sql.py:151]: I think so, the refactor on this operator is mainly to reduce duplicated code, and the logic, interface, and signature don't change in a backward incompatible one. But I don't know enough about how providers are released and how the core may affect it. cc @potiuk (sorry to ping you again..)

**Review by kaxil — APPROVED** [review]

File: airflow/operators/sql.py (hunk 13 above)
Comment (kaxil) [reply airflow/operators/sql.py:151]: I think it should work

**Comment by xinbinhuang** [comment]
@turbaszek @potiuk @kaxil Thank you for reviewing!
